### PR TITLE
472 sir scalars

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for 1st review from either @sergisiso or @arporter

**Comment by rupertford** [comment]
Taking this back temporarily as I forgot to update the if example in examples/nemo/eg4. I also see that the PR needs updating so I'll do this as well.

**Comment by rupertford** [comment]
Ready for review again.

File: src/psyclone/psyir/backend/sir.py [hunk 1]
```diff
@@ -117,6 +117,13 @@ def __init__(self, skip_nodes=False, indent_string="  ",
         # found in the PSyIR. This is required as the SIR declares
         # field names after the computation.
         self._field_names = set()
+        # The _scalar_names variable stores the unique scalar names
+        # found in the PSyIR. The current assumption is that scalars
+        # are temporaries. This is not necessarily correct and this
+        # problem is captured in issue #521. Scalar temporaries can be
```
Comment (arporter) [inline src/psyclone/psyir/backend/sir.py]: "captured" is a strong word - #521 contains no description :-) Please could you add one to that ticket.

File: src/psyclone/psyir/backend/sir.py [hunk 2]
```diff
@@ -345,7 +359,15 @@ def reference_node(self, node):
...
-        return "{0}make_var_access_expr(\"{1}\")".format(self._nindent,
+        # _scalar_names is a set so duplicates will be ignored. It
+        # captures all unique scalar names as scalars are currently
+        # treated as temporaries (#521 captures this). The simplest
+        # way to declare a scalar temporary in Dawn is to treat it as
+        # a field temporary (as the Dawn backend works out if a scalar
+        # is required).
+        self._scalar_names.add(node.name)
+
+        return "{0}make_field_access_expr(\"{1}\")".format(self._nindent,
```
Comment (arporter) [inline src/psyclone/psyir/backend/sir.py]: pylint complains about the indentation of the next line.

File: src/psyclone/psyir/backend/sir.py [hunk 3]
```diff
@@ -345,7 +359,15 @@ def reference_node(self, node):
             raise VisitorError(
                 "Method reference_node in class SIRWriter: SIR Reference "
                 "node is not expected to have any children.")
-        return "{0}make_var_access_expr(\"{1}\")".format(self._nindent,
+        # _scalar_names is a set so duplicates will be ignored. It
```
Comment (arporter) [inline src/psyclone/psyir/backend/sir.py]: So does `make_var_access_expr` not existing in the SIR API?

File: src/psyclone/psyir/backend/sir.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py]: Oops. Fixed.

File: src/psyclone/psyir/backend/sir.py (hunk 3 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py]: Yes it does exist, we just don't make use of it at the moment. The reason for this is that it is unclear how to declare scalars in the SIR and the proposed solution (from the Dawn developers) was to specify these as temporary fields (which is easy) as the Dawn backend translates temporary fields to the lowest possible dimension i.e. a scalar in our case.

File: src/psyclone/psyir/backend/sir.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py]: Point of order. It's an issue ;-)

Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py]: I've added some words to the issue. Hopefully they will be OK.

**Comment by rupertford** [comment]
Ready for next review from @arporter subject to travis being happy.

File: src/psyclone/psyir/backend/sir.py (hunk 3 above)
Comment (arporter) [reply src/psyclone/psyir/backend/sir.py]: OK. Thanks for explaining.

**Comment by rupertford** [comment]
Brought branch up-to-date with master and described the implementation of the support for scalars in the developer documentation. Ready for next review from @arporter.

**Pull request merged**